File: 1]. DSA + CP/1]. DSA/1]. Data Structures/01]. Array/C++/_039)_Find_SubArray_of_Given_Sum.cpp
#include<iostream>
using namespace std;

int subArraySum(int arr[], int n, int given_sum)
{
    int curr_sum = arr[0], start=0;
    for(int end=1; end<n; end++)
    {
        while(curr_sum > given_sum && start < end-1)
        {
            curr_sum -= arr[start];
            start++;
        }
        if(curr_sum == given_sum)
        {
            cout<<"Sum found between indexes "<< start <<" and "<< end-1 <<endl;
            return true;
        }
        if(end<n)
            curr_sum += arr[end];
    }
    cout<<"No subArray found"<<endl;
}

int main()
{
    int arr[] = {1, 4, 20, 3, 10, 5}, n=6, given_sum=23;
    subArraySum(arr, n, given_sum);
}

File: 1]. DSA + CP/1]. DSA/1]. Data Structures/01]. Array/Dart/_039)_Find_SubArray_of_Given_Sum.dart
void subArraySum(List<int> arr, int reqSum) {
  int sum = arr[0];
  int start = 0;
  for (int end = 1; end < arr.length; end++) {
    while (sum > reqSum && start < end - 1) {
      sum -= arr[start];
      start++;
    }
    if (sum == reqSum) {
      print("Subarray found.");
      print("Start: $start End: ${end - 1}");
      return;
    }
    if (end < arr.length) sum += arr[end];
  }

  print("Subarray not found.");
}

void main() {
  var arr = [1, 4, 20, 3, 10, 5];
  int sum = 23;
  subArraySum(arr, sum);
}
